lang/cs/Org.Apache.REEF.Examples/DriverRestart/HelloRestartDriver.cs

```diff
@@ -0,0 +1,181 @@
+    /// Once all tasks are running, the driver kills itself and expects the RM to restart it.
+    /// On restart, it expects all 5 of the running tasks to report back to it.
+    /// </summary>
+    public sealed class HelloRestartDriver : IObserver<IDriverRestartCompleted>, IObserver<IAllocatedEvaluator>, IObserver<IDriverStarted>, 
+        IObserver<IDriverRestarted>, IObserver<IActiveContext>, IObserver<IRunningTask>, IObserver<ICompletedTask>, IObserver<IFailedTask>
+    {
+        private static readonly Logger _Logger = Logger.GetLogger(typeof(HelloRestartDriver));
+        private const int AmountOfTasksToSubmit = 5;
+
+        private readonly IEvaluatorRequestor _evaluatorRequestor;
+        private readonly object _lockObj;
+
+        private int _runningTaskCount;
+        private int _finishedTaskCount;
+        private bool _restarted;
+
+        [Inject]
+        private HelloRestartDriver(IEvaluatorRequestor evaluatorRequestor)
+        {
+            _finishedTaskCount = 0;
+            _runningTaskCount = 0;
+            _evaluatorRequestor = evaluatorRequestor;
+            _restarted = false;
+            _lockObj = new object();
+        }
+
+        /// <summary>
+        /// This constructor is for the configuration to know that the driver has a restart
+        /// handler bound for it to enable restart, and is not intended for actual instantiation.
+        /// </summary>
+        [Inject]
+        private HelloRestartDriver()
+            : this(null)
+        {
+        }
+
+        /// <summary>
+        /// Submits the HelloRestartTask to the Evaluator.
+        /// </summary>
+        public void OnNext(IAllocatedEvaluator allocatedEvaluator)
+        {
+            var taskConfiguration = TaskConfiguration.ConfigurationModule
+                .Set(TaskConfiguration.Identifier, "HelloRestartTask")
+                .Set(TaskConfiguration.Task, GenericType<HelloRestartTask>.Class)
+                .Set(TaskConfiguration.OnMessage, GenericType<HelloRestartTask>.Class)
+                .Build();
+            allocatedEvaluator.SubmitTask(taskConfiguration);
+        }
+
+        /// <summary>
+        /// Called to start the driver.
+        /// </summary>
+        public void OnNext(IDriverStarted driverStarted)
+        {
+            _Logger.Log(Level.Info, string.Format("HelloRestartDriver started at {0}", driverStarted.StartTime));
+            for (var i = 0; i < AmountOfTasksToSubmit; i++)
+            {
+                this._evaluatorRequestor.Submit(new EvaluatorRequest(1, 64));
+            }
+        }
+
+        /// <summary>
+        /// Prints a restart message and enters the restart codepath.
+        /// </summary>
+        public void OnNext(IDriverRestarted value)
+        {
+            this._restarted = true;
+            _Logger.Log(Level.Info, "Hello! HelloRestartDriver has restarted!");
+        }
+
+        public void OnNext(IActiveContext value)
```

Comment:
This should never be called, right? The Driver submits a task to each of them, directly submits the task.